Add token type ids to CodeGenTokenizer

st81 opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes #28098 

## Who can review?

@ArthurZucker @younesbelkada

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @pacman100

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc and @younesbelkada

Documentation: @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

st81 committed:
Add create token type ids to CodeGenTokenizer

st81 commented:
This PR is WIP because I will check documentation update is needed and CodeGenTokenizer output includes `token_type_ids`.

st81 committed:
Fix inconsistent length of token type ids

st81 committed:
Format source codes

st81 committed:
Fix inconsistent order of methods

st81 commented:
@ArthurZucker @younesbelkada
I've made the necessary changes to address the issue. Could someone please review the updated pull request at your earliest convenience? Your feedback is greatly appreciated.

Thank you!

st81 committed:
Update docstring

ArthurZucker reviewed (commented):
Thanks! You might have to update the `test_tokenizer_integration` as it will now return token_type_ids. Let's make sure slow tokenizer tests pass as well!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_29265). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

st81 commented:
> Thanks! You might have to update the `test_tokenizer_integration` as it will now return token_type_ids. Let's make sure slow tokenizer tests pass as well!

@ArthurZucker 

Thank you for taking the time to review the pull request! It seems that `test_tokenizer_integration` is missing from `tests/models/codegen/test_tokenization_codegen.py`. Would you like me to include `test_tokenizer_integration` similar to the other tokenizers tests?

st81 committed:
add test_tokenizer_integration test

st81 committed:
Format source codes

st81 commented:
> Thanks! You might have to update the `test_tokenizer_integration` as it will now return token_type_ids. Let's make sure slow tokenizer tests pass as well!

@ArthurZucker 
It seems that `test_tokenizer_integration` is missing from `tests/models/codegen/test_tokenization_codegen.py`. So I’ve added `test_tokenizer_integration`. Could you review again?

ArthurZucker commented:
Will review tomorrow sorry !

ArthurZucker commented on src/transformers/models/codegen/tokenization_codegen_fast.py:
```diff
@@ -177,6 +176,35 @@ def _encode_plus(self, *args, **kwargs) -> BatchEncoding:
 
         return super()._encode_plus(*args, **kwargs)
 
+    def create_token_type_ids_from_sequences(
```
let's use copied from here! (from the slow code)

st81 replied:
I have added `Copied from ...` comment!
(Please let me know if this change is not what you intended)

ArthurZucker replied:
yes! 🤗

ArthurZucker reviewed (commented):
LGTM just a small nit

st81 committed:
Add `copied from` comment to CodeGenTokenizerFast

st81 committed:
Add doc of create_token_type_ids_from_sequences

st81 commented:
I've addressed your comment and also added documentation for `CodeGenTokenizer.create_token_type_ids_from_sequences` ! Could you please review it again?

ArthurZucker commented on tests/models/codegen/test_tokenization_codegen.py:
```diff
@@ -263,3 +263,10 @@ def test_truncation(self):
+        # fmt: off
+        expected_encoding = {'input_ids': [[41762, 364, 357, 36234, 1900, 355, 12972, 13165, 354, 12, 35636, 364, 290, 12972, 13165, 354, 12, 5310, 13363, 12, 4835, 8, 3769, 2276, 12, 29983, 45619, 357, 13246, 51, 11, 402, 11571, 12, 17, 11, 5564, 13246, 38586, 11, 16276, 44, 11, 4307, 346, 33, 861, 11, 16276, 7934, 23029, 329, 12068, 15417, 28491, 357, 32572, 52, 8, 290, 12068, 15417, 16588, 357 ...
+        # fmt: on
+        self.tokenizer_integration_test_util(expected_encoding, "Salesforce/codegen-350M-mono", padding=False)
```
```suggestion
        self.tokenizer_integration_test_util(expected_encoding, "Salesforce/codegen-350M-mono", padding=False, return_token_type_ids=True)
```
for BC we should make sure that `return_token_type_ids` is set to False by default

st81 replied:
I've updated this test as you suggested.

ArthurZucker reviewed (commented):
Currently with the state of this PR:

```python 
In [1]:  from transformers import AutoTokenizer

In [2]: tokenizer = AutoTokenizer.from_pretrained("Salesforce/codegen-350M-mono")

In [3]:  tokenizer("Hey how are you")
Out[3]: {'input_ids': [10814, 703, 389, 345], 'token_type_ids': [0, 0, 0, 0], 'attention_mask': [1, 1, 1, 1]}

```
this is not backward compatible. `return_token_type_ids` should be set to false

st81 committed:
Make return_token_type_ids False by default

ArthurZucker commented on tests/models/codegen/test_tokenization_codegen.py:
```diff
@@ -263,3 +263,12 @@ def test_truncation(self):
     def test_padding_different_model_input_name(self):
         pass
+
+    # @slow
```
```suggestion
    @slow
```

st81 replied:
Sorry I forgot to uncomment it. I've fixed it!

ArthurZucker reviewed (commented):
We have to change the default `return_token_type_ids` not in the test (it's alright to keep it) but most importantly in the __init__ of both tokenizers, set it to `        self.return_token_type_ids = return_token_type_ids` and `return_token_type_ids=False` in the args!

st81 committed:
Make test_tokenizer_integration as slow test

st81 commented:
@ArthurZucker 

Could you take a look at [this comment](https://github.com/huggingface/transformers/pull/29265#issuecomment-1983580444)? Your feedback is greatly appreciated.

st81 commented:
> @st81 Just to let you know @ArthurZucker is off for a week

Thank you for letting me know about that! I'll wait for him to return.

st81 commented:
@ArthurZucker 

> We have to change the default `return_token_type_ids` not in the test (it's alright to keep it) but most importantly in the **init** of both tokenizers, set it to ` self.return_token_type_ids = return_token_type_ids` and `return_token_type_ids=False` in the args!

I think current changes are fine to make the default `return_token_type_ids` is `False` because [it looks like the default value of `return_token_type_ids` is controlled by whether `self.model_input_names` contains `"token_type_ids"` or not](https://github.com/huggingface/transformers/blob/8dd4ce6f2cea1316cb5bd5ccbd348310ced17856/src/transformers/tokenization_utils_base.py#L3494-L3495). What are your thoughts on this?

ArthurZucker reviewed (commented):
Hey, regarding: 
> When looking at other tokenizers in the repository, I found that none of them have return_token_type_ids as an argument in the __init__ function, and set return_token_type_ids as a class variable of the tokenizer like self.return_token_type_ids = return_token_type_ids. Would it be problematic from the perspective of library consistency to make CodeGenTokenizer the only one with such a specification?

it is not problematic no, as this is just a means to make sure we respect backward compatibility. 

Let's set it to False for this model only.

st81 committed:
Add return_token_type_ids to tokenizer init arg

st81 committed:
Add test for tokenizer's init return_token_type_ids

st81 committed:
Format source codes

st81 commented:
> We have to change the default `return_token_type_ids` not in the test (it's alright to keep it) but most importantly in the **init** of both tokenizers, set it to ` self.return_token_type_ids = return_token_type_ids` and `return_token_type_ids=False` in the args!

I've made changes based on your [comment](https://github.com/huggingface/transformers/pull/29265#pullrequestreview-1921451566). Could you take a look?

ArthurZucker commented on src/transformers/models/codegen/tokenization_codegen.py:
```diff
@@ -150,13 +152,17 @@ def __init__(
         pad_token = AddedToken(pad_token, special=True) if isinstance(pad_token, str) else pad_token
         self.add_bos_token = add_bos_token
+        self.return_token_type_ids = return_token_type_ids
+        if self.return_token_type_ids:
+            self.model_input_names.append("token_type_ids")
```
if we always add it does it make a difference? It's my first time seeing this fix, so a bit suprised that we have to do that

st81 replied:
Yes. This actually controls whether the tokenizer returns `token_type_ids` or not in [here](https://github.com/huggingface/transformers/blob/8dd4ce6f2cea1316cb5bd5ccbd348310ced17856/src/transformers/tokenization_utils_base.py#L3494-L3495). The tokenizer doesn't return `token_type_ids` without this line.

ArthurZucker replied:
Sorry you are right! It's not really well made. I'll approve, but would be nice if we can fix self.model_iinput_names based on the value of `return_token_type_ids`. For now only the reverse is True.

ArthurZucker commented on src/transformers/models/codegen/tokenization_codegen.py:
```diff
@@ -150,13 +152,17 @@ def __init__(
         pad_token=None,
         add_prefix_space=False,
         add_bos_token=False,
+        return_token_type_ids=False,
```
that is the expected change

ArthurZucker reviewed (commented):
LGTM otherwise

ArthurZucker reviewed (approved):
Thanks

ArthurZucker commented:
Sorry for the delay, LGTM

st81 commented:
Thank you so much for reviewing!
